prometheus/promhttp/instrument_server.go

```diff
@@ -393,8 +398,11 @@ func isLabelCurried(c prometheus.Collector, label string) bool {
 // unnecessary allocations on each request.
 var emptyLabels = prometheus.Labels{}
 
-func labels(code, method bool, reqMethod string, status int, extraMethods ...string) prometheus.Labels {
+func labels(code, method, hasExtraLabelsCtx bool, reqMethod string, status int, extraMethods ...string) prometheus.Labels {
 	if !(code || method) {
+		if hasExtraLabelsCtx {
+			return prometheus.Labels{}
+		}
 		return emptyLabels
```

Comment:
Thank you @bwplotka 

I removed the usage of the global `emptyLabels` map because as you say it's highly misleading and it makes it trivial to introduce new data races by writing into the output of the `labels` func.